// testapi2/dist/tags/DocUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toJsonSchema = exports.isGotData = exports.mergeData = void 0;
const lodash_1 = require("lodash");
const convert = require('to-json-schema');
function mergeData(data) {
    if (Array.isArray(data)) {
        if (data.length === 0)
            return [];
        let rs = mergeData(data[0]);
        for (let i = 1; i < data.length; i++) {
            lodash_1.merge(rs, mergeData(data[i]));
        }
        return [rs];
    }
    else if (typeof data === 'object' && data !== null) {
        const rs = {};
        for (let k in data) {
            rs[k] = mergeData(data[k]);
        }
        return rs;
    }
    return data;
}
exports.mergeData = mergeData;
function isGotData(obj, isCheckEmptyObject = true) {
    if (obj === null || obj === undefined || obj === '')
        return false;
    if (Array.isArray(obj))
        return true;
    if (typeof obj === 'object') {
        if (isCheckEmptyObject)
            return Object.keys(obj).length > 0;
        return !!obj;
    }
    return false;
}
exports.isGotData = isGotData;
function toJsonSchema(data, opts = {
    objects: { additionalProperties: false },
    arrays: { mode: 'first' },
    strings: { detectFormat: false },
}) {
    return convert(mergeData(data), opts);
}
exports.toJsonSchema = toJsonSchema;
