// testapi2/dist/tags/Utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const DocUtils_1 = require("./DocUtils");
const Tag_1 = require("./Tag");
const global_config_1 = require("../global_config");
const crypto_1 = require("crypto");
Object.assign(global_config_1.globalConfig.utils, {
    schema(obj, opts) {
        return DocUtils_1.toJsonSchema(obj, opts);
    },
    base64(txt) {
        return txt && Buffer.from(txt).toString('base64');
    },
    md5(txt) {
        return txt && crypto_1.createHash('md5').update(txt).digest('hex');
    },
    sign(vl) {
        vl = +vl;
        return (vl > 0 ? '+' : '') + vl;
    }
});
class Utils extends Tag_1.Tag {
    constructor(attrs) {
        super(attrs);
        const { utils } = global_config_1.globalConfig;
        eval(`utils['${this.func}'] = ${this.script}`);
    }
    exec() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
}
exports.Utils = Utils;
