// testapi2/dist/tags/yaml/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.head = exports.del = exports.patch = exports.put = exports.post = exports.get = exports.api = void 0;
const js_yaml_1 = require("js-yaml");
const Api_1 = require("../Api");
exports.api = new js_yaml_1.Type('!api', {
    kind: 'mapping',
    instanceOf: Api_1.Api,
    construct: attrs => new Api_1.Api(attrs)
});
exports.get = new js_yaml_1.Type('!get', {
    kind: 'mapping',
    instanceOf: Api_1.Api,
    construct: attrs => new Api_1.Api(Object.assign(Object.assign({}, attrs), { method: Api_1.Method.GET }))
});
exports.post = new js_yaml_1.Type('!post', {
    kind: 'mapping',
    instanceOf: Api_1.Api,
    construct: attrs => new Api_1.Api(Object.assign(Object.assign({}, attrs), { method: Api_1.Method.POST }))
});
exports.put = new js_yaml_1.Type('!put', {
    kind: 'mapping',
    instanceOf: Api_1.Api,
    construct: attrs => new Api_1.Api(Object.assign(Object.assign({}, attrs), { method: Api_1.Method.PUT }))
});
exports.patch = new js_yaml_1.Type('!patch', {
    kind: 'mapping',
    instanceOf: Api_1.Api,
    construct: attrs => new Api_1.Api(Object.assign(Object.assign({}, attrs), { method: Api_1.Method.PATCH }))
});
exports.del = new js_yaml_1.Type('!delete', {
    kind: 'mapping',
    instanceOf: Api_1.Api,
    construct: attrs => new Api_1.Api(Object.assign(Object.assign({}, attrs), { method: Api_1.Method.DELETE }))
});
exports.head = new js_yaml_1.Type('!head', {
    kind: 'mapping',
    instanceOf: Api_1.Api,
    construct: attrs => new Api_1.Api(Object.assign(Object.assign({}, attrs), { method: Api_1.Method.HEAD }))
});
