// testapi2/dist/tags/yaml/file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fileContent = exports.fileStream = void 0;
const js_yaml_1 = require("js-yaml");
const path_1 = require("path");
const fs_1 = require("fs");
const Testcase_1 = require("../Testcase");
const tty_1 = require("tty");
const __1 = require("..");
function loadContent(file) {
    const cnt = fs_1.readFileSync(file, 'utf8').toString();
    if (/\.ya?ml$/.test(file))
        return js_yaml_1.safeLoad(cnt, { schema: __1.SCHEMA });
    if (/\.json$/.test(file))
        return JSON.parse(cnt);
    if (/\.csv$/.test(file))
        return cnt.split('\n').map(e => e.split(','));
    return cnt;
}
exports.fileStream = new js_yaml_1.Type('!upload', {
    kind: 'scalar',
    instanceOf: tty_1.ReadStream,
    construct: (data) => {
        const file = path_1.join(Testcase_1.Testcase.RootDir, data);
        return fs_1.createReadStream(file);
    }
});
exports.fileContent = new js_yaml_1.Type('!file', {
    kind: 'scalar',
    instanceOf: String,
    construct: (data) => {
        const file = path_1.join(Testcase_1.Testcase.RootDir, data);
        return loadContent(file);
    }
});
