// testapi2/dist/tags/yaml/validate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validate = exports.status = exports.statusText = void 0;
const js_yaml_1 = require("js-yaml");
const Validate_1 = require("../Validate");
exports.statusText = new js_yaml_1.Type('!status-text', {
    kind: 'scalar',
    construct: (value) => {
        return new Validate_1.Validate({
            title: 'Response status text',
            func: 'match',
            args: [
                '${$.response.statusText}',
                `${value}`
            ]
        });
    }
});
exports.status = new js_yaml_1.Type('!status', {
    kind: 'scalar',
    construct: (value) => {
        if (typeof value === 'string') {
            value = value.split(',').map(e => +e);
        }
        else {
            value = [value];
        }
        return new Validate_1.Validate({
            title: 'Response status',
            func: 'in',
            args: [
                '${$.response.status}',
                value
            ]
        });
    }
});
exports.validate = new js_yaml_1.Type('!validate', {
    kind: 'mapping',
    instanceOf: Validate_1.Validate,
    construct: (attrs) => new Validate_1.Validate(attrs)
});
